File: code/python/337.py
class Solution:
    def rob(self, root: TreeNode) -> int:
        #memo = dict()
        def robTree(root):
            if root is None:
                return 0,0
            left,leftNon = robTree(root.left)
            right,rightNon = robTree(root.right)
            res = leftNon+rightNon+root.val
            resNon = max(left,leftNon)+max(rightNon,right)
            return res,resNon
        res,resNon = robTree(root)
        return max(res,resNon)

File: code/java/337.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int[] robTree(TreeNode root){
        if(root==null) return new int[]{0,0};
        int[] left = robTree(root.left);
        int[] right = robTree(root.right);
        int res = left[1]+right[1]+root.val;
        int resNon = Math.max(left[0],left[1])+Math.max(right[0],right[1]);
        return new int[]{res,resNon};
    }

    public int rob(TreeNode root) {
        int[] res = robTree(root);
        return Math.max(res[0],res[1]);
    }
}